File: ObjC/SIALogger/SIALogger/Sources/Colored/SIALogColoredFormatter.m
//
//  SIALogColoredFormatter.m
//  SIALogger
//
//  Created by Alexander Ivlev on 06/06/16.
//  Copyright © 2016 Alexander Ivlev. All rights reserved.
//

#import "SIALogColoredFormatter.h"
#import "SIALogFormatter.h"

@interface SIALogColoredFormatter ()

@property (nonatomic, strong) NSArray* formatters;

@end

@implementation SIALogColoredFormatter

- (instancetype)initWithFormat:(NSString*)format {
  assert(nil != format);
  self = [super init];
  if (self) {
    self.formatters = [SIALogColoredFormatter parse:format];
  }
  
  return self;
}

- (NSString*)toString:(SIALogMessage*)msg WithColoredMethod:(SIALogColoredMethod)colored {
  NSMutableString* result = [NSMutableString string];
  
  size_t index = 0;
  for (SIALogFormatter* formatter in self.formatters) {
    if (0 == index%2) {
      [result appendString:[formatter toString:msg]];
    } else {
      [result appendString:colored([formatter toString:msg])];
    }
    index++;
  }
  
  return [result copy];
}

+ (NSArray*)parse:(NSString*)format {
  NSMutableArray* result = [NSMutableArray array];
  
  NSString* formatIter = [format copy];
  
  do {
    NSRange range = [formatIter rangeOfString:@"%c"];
    if (NSNotFound != range.location) {
      id formatter = [[SIALogFormatter alloc] initWithFormat:[formatIter substringToIndex:range.location]];
      [result addObject:formatter];
      
      formatIter = [formatIter substringFromIndex:range.location+range.length];
      continue;
    }
    
    if (0 < formatIter.length) {
      [result addObject:[[SIALogFormatter alloc] initWithFormat:formatIter]];
      break;
    }
    
  } while(true);
  
  return [result copy];
}

@end

File: Swift/SIALogger/SIALogger/Colored/SIALogColoredFormatter.swift
//
//  SIALogColoredFormatter.swift
//  SIALogger
//
//  Created by Alexander Ivlev on 06/06/16.
//  Copyright © 2016 Alexander Ivlev. All rights reserved.
//

public class SIALogColoredFormatter {
  public let format: String
  
  public required init(format: String) {
    self.format = format
    self.formatters = SIALogColoredFormatter.parse(format)
  }
  
  public func toString(message: SIALogMessage, coloredMethod: (String) -> String) -> String {
    var result = ""
    
    var index = 0
    for formatter in self.formatters {
      if 0 == index%2 {
        result += formatter.toString(message)
      } else {
        result += coloredMethod(formatter.toString(message))
      }
      
      index += 1
    }
    
    return result
  }
  
  private let formatters : [SIALogFormatter]
  
  private static func parse(format: String) -> [SIALogFormatter] {
    var result : [SIALogFormatter] = []

    var formatIter = format
    
    while let range = formatIter.rangeOfString("%c") {
      result.append(SIALogFormatter(format: formatIter.substringToIndex(range.startIndex)))
      
      formatIter = formatIter.substringFromIndex(range.endIndex);
    }
    
    if !formatIter.isEmpty {
      result.append(SIALogFormatter(format: formatIter))
    }
    
    return result
  }
  
}
